Parse FDQ extensions in isa string as in spec

marcfedorow opened the pull request:
See: 27.11 Subset Naming Convention (Table 27.1)

marcfedorow commented:
@aswaterman, actually spike parses isa improperly:
- it allows wrong letter order (e.g. imcaqdf) 
- it does not allow no underscore before first zshx subext (except bitmanip)
- it does not allow underscore before bitmanip

Are you or if I make this parsing consistent with the latest riscv-isa-spec?

aswaterman commented:
Sure, feel free to append to this PR.

aswaterman commented on riscv/processor.cc:
```diff
@@ -248,6 +248,11 @@ void processor_t::parse_isa_string(const char* str)
+      switch (*p) {
+        case 'q': max_isa |= 1L << ('d' - 'a');
+        case 'd': max_isa |= 1L << ('f' - 'a');
+        // case 'f': extension_table[EXT_ZICSR] = true;
```
Please don't include the commented-out code.

Note that Spike _necessarily_ has Zicsr, because Zicsr is implied by the privileged architecture.  We don't need to call it out specially.  I suppose we could accept the string and ignore it, but we shouldn't require the string.

marcfedorow commented:
@aswaterman I tried to achieve some equilibrium between not making this parser as complicated as in gcc but setting required misa bits on the other hand.
Not sure if I missed something spike-specific.

aswaterman commented on riscv/processor.cc:
```diff
@@ -217,144 +217,120 @@ void processor_t::parse_priv_string(const char* str)
+    if (!*all_subsets)
+      bad_isa_string(str, "Wrong order");
+    switch (*p) {
+      case 'q': max_isa |= 1L << ('d' - 'a');
```
This needs to be `(1L << ('d' - 'a')) | (1L << ('f' - 'a'))`, because Q implies D then D implies Q.  I'll fix it as part of testing it out, though.

aswaterman replied:
Nevermind, missed that this is a fallthrough...  good ol' C.

aswaterman commented:
Looks like this still has some bugs: e.g., `--isa=rv64gc` doesn't work.

aswaterman commented:
Also, custom extensions are partially broken.  `--isa=rv64gc_xfoobar` should print `couldn't find extension 'foobar' in shared library 'libcustomext.so'`

`xbitmanip` still seems to be working correctly.

As a special case `xdummy` is simply supposed to be ignored.  This seems to be working correctly, but make sure it still works after fixing the above.

marcfedorow commented:
@aswaterman, I think it's ready for review
- fixed index for 'g' letter
- reverted remove of `register_extension` call

Actually `rv64gc_xfoobar` never worked (while `rv64gcxfoobar` was). Currently both of them do the same thing.

aswaterman commented:
Looks good.  If any other issues arise, we can deal with them separately.
